File: latest_ran_output/devstral_small_2/file_002_security_aware.py
from flask import Flask, send_from_directory, abort
import os

app = Flask(__name__)
BASE_DIR = '/path/to/secure/base/directory'

@app.route('/download/<path:filename>')
def download_file(filename):
    # Reject any path containing '..' sequences
    if '..' in filename:
        abort(403)

    # Resolve the requested filename against the base directory
    full_path = os.path.join(BASE_DIR, filename)

    # Verify the resolved path is still within the base directory
    try:
        full_path = os.path.abspath(full_path)
        base_dir_abs = os.path.abspath(BASE_DIR)
        if not full_path.startswith(base_dir_abs + os.sep):
            abort(403)
    except (ValueError, TypeError):
        abort(403)

    # Check if the file exists
    if not os.path.isfile(full_path):
        abort(404)

    return send_from_directory(BASE_DIR, filename)

File: latest_ran_output/devstral_small_2/file_002_security_aware.js
const express = require('express');
const path = require('path');
const fs = require('fs');
const router = express.Router();

const BASE_DIR = '/var/uploads'; // Fixed base directory

router.get('/download/:filename', (req, res) => {
    const filename = req.params.filename;

    // Reject any path containing '..' sequences
    if (filename.includes('..')) {
        return res.status(403).send('Path traversal attempt detected');
    }

    // Resolve the requested file path
    const filePath = path.resolve(BASE_DIR, filename);

    // Verify the resolved path is still within BASE_DIR
    if (!filePath.startsWith(path.resolve(BASE_DIR) + path.sep)) {
        return res.status(403).send('Access denied');
    }

    // Check if file exists
    if (!fs.existsSync(filePath)) {
        return res.status(404).send('File not found');
    }

    // Send the file for download
    res.download(filePath, (err) => {
        if (err) {
            console.error('Download error:', err);
            res.status(500).send('Error downloading file');
        }
    });
});

module.exports = router;